advection_test/functions.py

```diff
@@ -0,0 +1,1302 @@
+                )
+
+    if grid == "D":
+        for jj in range(
+            dimensions["nhalo"] - 1, dimensions["ny"] + dimensions["nhalo"] + 1
+        ):
+            for ii in range(
+                dimensions["nhalo"] - 1, dimensions["nx"] + dimensions["nhalo"] + 1
+            ):
+                dist = dx.data[ii, jj]
+                v_grid[ii, jj] = (
+                    0 if dist == 0 else (psi.data[ii, jj] - psi.data[ii - 1, jj]) / dist
+                )
+
+                dist = dy.data[ii, jj]
+                u_grid[ii, jj] = (
+                    0
+                    if dist == 0
+                    else -1.0 * (psi.data[ii, jj] - psi.data[ii, jj - 1]) / dist
+                )
+
+    return u_grid, v_grid
+
+
+def create_initialState_testCase1a(
+    grid_data, dimensions, units, origins, backend, smoke_dict, pressure_dict
+):
+    """
+    Use: initialState =
+    create_initialState1a(grid_data, dimensions, units, origins,
+                        backend, smoke_dict, pressure_dict)
+
+    Creates inital state from the fortran test_case 1 streamfunction configuration -
+    pressure, gaussian smoke distribution, u and v winds on C-grid.
+
+    Inputs:
+    - grid_data: configuration['grid_data']
+    - dimensions: Dict{'nx', 'ny', 'nx1', 'ny1', 'nxhalo', 'nyhalo', 'tile'}
+    - units: Dict{'coord', 'dist', 'mass', 'psi', 'wind'}
+    - origins: Dict{'halo', 'compute_2d'}
+    - backend: 'numpy'
+    - smoke_dict: Dict{'center_tile', 'rank', 'smoke_base'}
+    - pressure_dict: Dict{'pressure_base'}
+
+    Outputs:
+    - initialState: Dict{'delp', 'smoke', 'uc', 'vc'} - 3D with a
+        single layer in the vertical
+    """
+    lonA_halo = Quantity(
+        grid_data.lon_agrid.data,
+        ("x_halo", "y_halo"),
+        units["coord"],
+        origins["halo"],
+        (dimensions["nxhalo"], dimensions["nyhalo"]),
+        backend,
+    )
+    latA_halo = Quantity(
+        grid_data.lat_agrid.data,
+        ("x_halo", "y_halo"),
+        units["coord"],
+        origins["halo"],
+        (dimensions["nxhalo"], dimensions["nyhalo"]),
+        backend,
+    )
+
+    # SMOKE
+    gaussian_multiplier = create_gaussianMultiplier(
+        lonA_halo.data,
+        latA_halo.data,
+        dimensions,
+        center_tile=smoke_dict["center_tile"],
+        mpi_rank=smoke_dict["rank"],
+    )
+    smoke = Quantity(
+        gaussian_multiplier * smoke_dict["smoke_base"],
+        ("x", "y"),
+        units["mass"],
+        origins["compute_2d"],
+        (dimensions["nx"], dimensions["ny"]),
+        backend,
+    )
+
+    # PRESSURE
+    delp = Quantity(
+        np.ones(smoke.data.shape) * pressure_dict["pressure_base"],
+        ("x", "y"),
+        units["pressure"],
+        origins["compute_2d"],
+        (dimensions["nx"], dimensions["ny"]),
+        backend,
+    )
+
+    # STREAMFUNCTION
+    _, psi_staggered = calculate_streamfunction_testCase1a(
```

Comment:
why does this function return psi?

Reply:
psi is needed if we want winds on A-grid or D-grid. Since those are options for the calculate_winds function, psi is also returned. But since the initialization only needs winds on C-grid, we skip storing it here.